### PR TITLE
Add backend exception handling and Un_cps design docs

### PR DESCRIPTION
**Pull request by mshinwell**
This is just an import from the old `flambda2.0-stable` branch with a PR created for visibility, review/editing will come later.  I've renamed `un_cps.md` to `to_cmm.md`.

### CODE REVIEW

File: middle_end/flambda2/docs/to_cmm.md [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+reference to store that information during translation). This static result
+(defined in `un_cps_result.ml`) will be grown by the translation, by either
+`let_symbol` bindings or sets of closures bindings. Translations of flambda2
+expression that can result in pre-allocation of static constants go through
+the `un_cps_static.ml` file.
+
+### Tagging and unboxing
+
+In the code, this is one of the rather easy points. Flambda2 has a notion
+of kinds, which offser enough information about the memory representation
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:223]:
```suggestion
of kinds, which offers enough information about the memory representation
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+a tagged integer, and which deal with untagged integers), this offers a
+rather comprehensive way for `un_cps` to know when or where to tag/untag
+things. See the following for more details:
+- `middle_end/flambda/terms/flambda_primitive.ml` for the data about kinds
+  consumed/returned by flambda primitives (particularly the
+  `arg_kind_of_*_primitive` and `result_kind_of_*_primitive`)
+- [the cmm doc](cmm.md) for info about the primitives in cmm
+  (this is still a TODO)
+
+Unboxing is also straightforward: as mentioneed earlier, flambda2 already
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:236]:
```suggestion
Unboxing is also straightforward: as mentioned earlier, flambda2 already
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+### "Big" and "small" values
+
+We'll describe here the way "big" and "small" values are translated, since
+the rest is relatively simple and can be understood directly from the code.
+
+#### Int64 on 32-bit archs
+
+Since in most cases, these are handled by C functions, the C convention is used
+throughtou all the code: on a 32-bit arch, an Int64 is split into its lower
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:252]:
```suggestion
throughout all the code: on a 32-bit arch, an Int64 is split into its lower
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+is to check the potential trap actions that can be associated to a continuation
+call. But again, these are relatively simple to translate since cmm now has
+a trap mechanism to mirror that of flambda2.
+
+### Let-binding substitution
+
+This is the most complex part of `un_cps` and also the part most sensitive
+to changes (even innocuous changes in `un_cps` might break becasue of this).
+The goal is to substitute let-bound variable's bodies during translation, because
+it allows for more optimizations. There are two technical challengs
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:304]:
```suggestion
it allows for more optimizations. There are two technical challenges
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+with this:
+- keep the correct order of evaluation (with respect to effects and co-effects)
+- yet still allow for some re-ordering of evaluation, such as pure computations
+  across other computations (as long as it doesn't cross into a recursive
+  continuations, in which case the expression might be valuated more than once).
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:309]:
```suggestion
  continuations, in which case the expression might be evaluated more than once).
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+Let-bindings are stored in the environment in a few structures:
+- pure bindings are stored in a dedicated map
+- all other bindings are stored in "stages". Intuitively a stage is a set of bindings
+  that can all be commuted with each other, while keeping the same semantics. This
+  means that a stage is either: a set of let-bindings which only have co-effects,
+  or a single effectful binding.
+These two structures are used to store enough information to perform substitution
+if needed. Additionally, another map is used to relate all flambda2 variables to
+a corresponding cmm variables. This mapping is used when `un_cps` decides not to
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:325]:
```suggestion
corresponding cmm variables. This mapping is used when `un_cps` decides not to
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+The points at which the env should be flushed can basically be understood as any
+operations that cannot be expressed as a let-binding, so any control flow,
+including non-inlined continuations, but excluding function calls (since these can
+be seen as effectful let-bindings).
+
+### Closure offsets
+
+Closure offset computation is done in two steps. First is to accumulate all
+information about sets of closures and members of these in ia mutable
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:364]:
```suggestion
information about sets of closures and members of these in a mutable
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+structure: by iterating over all sets of closure sin the flamdba2 program body,
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:365]:
```suggestion
structure: by iterating over all sets of closures in the flamdba2 program body,
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+we build a record for each set of closures. These records contains a list of
+closures and of env vars that belong to the set of closures; we will refer to
+them as slots. A slot is a member of a set of closures, currently a slot is
+occupied by either a closure or an env var.  Additionally, each slot record all
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:369]:
```suggestion
occupied by either a closure or an env var.  Additionally, each slot records all
```

File: middle_end/flambda2/docs/backend_exceptions.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+the base stack (and when the handler finishes execution flows to the same code
+as after the body, with the base stack).
+
+## Problematic cases
+
+This all works well as long as all code paths go through the `Lpushtrap` and
+`Lpoptrap` instructions in a well-parenthesized manner. However, this invariant
+can be broken by the static exceptions used in the compiler representations
+from Lambda and onwards. In particular, although it only occurs in specific
+circumstences, it is allowed to have inside the body of a `try...with`
```
Comment (xclerc) [inline middle_end/flambda2/docs/backend_exceptions.md:45]:
```suggestion
circumstances, it is allowed to have inside the body of a `try...with`
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+- cmm only operates on machine integers, so tagging explicitly has to be
+  translated in terms of arithmetic (left shift and addition for tagging for
+  instance)
+- allocations still benefit from some builtin handling, but block headers
+  are integers from cmm onwards, so concrete integer values for the headers
+  must be computed (whereas before, it was only the tag info)
+- field projection must be translated as a memory address computation
+  (pointer of the value + offset of the field, which can depend on the integer
+  size for the targeted architecture), and then a memory read
+- similarly, field assignements must be expressed as memory address
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:68]:
```suggestion
- similarly, field assignments must be expressed as memory address
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+flambda2.
+
+## Concrete details
+
+### Organization of the code and main control flow of the translation
+
+`un_cps` translate flambda2 expressions using two traversals of an flambda2
+body.
+
+The first traverssal iterates over all sets of closures defined by the body.
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:204]:
```suggestion
The first traversal iterates over all sets of closures defined by the body.
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+### "Big" and "small" values
+
+We'll describe here the way "big" and "small" values are translated, since
+the rest is relatively simple and can be understood directly from the code.
+
+#### Int64 on 32-bit archs
+
+Since in most cases, these are handled by C functions, the C convention is used
+throughtou all the code: on a 32-bit arch, an Int64 is split into its lower
+bits and its higher bits, and then carrried on two registers.  Handling these
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:253]:
```suggestion
bits and its higher bits, and then carried on two registers.  Handling these
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+[2]: this is done by the caller of `Env.bind_variable`, on variables that are used
+exactly once.
+
+This process allows to correctly substitute let-bound variable's bodies when
+adequate, but for variables not substituted, the let-binding information is
+"only" in the env: when translating the let-binding, all that is done is add the
+let-binding info in the env, thus if the variable's body is not substituted, it
+simply stays in the env. Hence, the env needs to be "flushed" of its accumulated
+let-bindings at some precise points to ensure that all let-bindings are correctly
+translated and present in the generated cmm code. This presents two challenge:
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:349]:
```suggestion
translated and present in the generated cmm code. This presents two challenges:
```

File: middle_end/flambda2/docs/to_cmm.md [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+operations that cannot be expressed as a let-binding, so any control flow,
+including non-inlined continuations, but excluding function calls (since these can
+be seen as effectful let-bindings).
+
+### Closure offsets
+
+Closure offset computation is done in two steps. First is to accumulate all
+information about sets of closures and members of these in ia mutable
+structure: by iterating over all sets of closure sin the flamdba2 program body,
+we build a record for each set of closures. These records contains a list of
```
Comment (xclerc) [inline middle_end/flambda2/docs/to_cmm.md:366]:
```suggestion
we build a record for each set of closures. These records contain a list of
```

File: middle_end/flambda2/docs/backend_exceptions.md [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+structure a sub-expression that raises a static exception whose handler is
+outside the body of the `try...with`. This looks like a simple detail, as
+the code that emits the traps simply needs to insert traps around these
+static raising constructs, but this is actually also important for earlier
+passes as some optimisations involving static exceptions that would be correct
+if both the handler and the raise are in the same exception context become
+incorrect if the contexts differ.
+
+To deal with this, the relevant optimisations in the compiler carry a `try_depth`
```
Comment (xclerc) [inline middle_end/flambda2/docs/backend_exceptions.md:54]: (Not sure it is worth mentioning, but another consequence is that
we need to generate multiple `Lpoptrap` in such a situation.)

File: middle_end/flambda2/docs/backend_exceptions.md [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+representations except the very last ones (linear and the generated assembly)
+have some kind of expression structure where the only way to change the trap
+context is to enter or exit a trywith block.
+
+But there have been some experiments (in particular the flambda2 project, but
+not only) that use a representation of the code that is either explicitly
+or implicitly a control-flow graph. For those representations, the ability
+to re-order the nodes (or instruction blocks) of the graph is an important
+property, and at least for flambda2 it's very easy to end up with a graph
+for which there is no straightforward transformation back into a structured
```
Comment (xclerc) [inline middle_end/flambda2/docs/backend_exceptions.md:83]: Out of curiosity, how important is this property?

Comment (lthls) [reply middle_end/flambda2/docs/backend_exceptions.md:83]: This was written to justify why we couldn't easily reuse the existing `Ctrywith`/`Itrywith` instruction in the backend. If we could rebuild structured expressions, we wouldn't have needed to introduce explicit traps so early in the backend.

Comment (xclerc) [reply middle_end/flambda2/docs/backend_exceptions.md:83]: I see, thanks; what I was actually wondering was along the lines
of "Once we have a CFG, do we need to go back to a _structured_
representation? it looks like at this point we only want to emit
code."

Comment (lthls) [reply middle_end/flambda2/docs/backend_exceptions.md:83]: I wasn't aware of the work on CFG representations for Linear instructions when I wrote this document. Of course we don't need to go back to a structured representation at that point, but the Flambda2 terms (which are also similar to control-flow graphs) need to go through the normal compilation pipeline (Cmm, Mach, and so on) which uses structured expressions.

Comment (xclerc) [reply middle_end/flambda2/docs/backend_exceptions.md:83]: Understood, thanks for the contextualisation.

File: middle_end/flambda2/docs/backend_exceptions.md (hunk 16 above)
Comment (mshinwell) [reply middle_end/flambda2/docs/backend_exceptions.md:55]: I've edited the doc to add a mention of multiple traps.

File: middle_end/flambda2/docs/to_cmm.md [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+  of closure that can be statically allocated
+- try and avoid using a global mutable state, but still try and re-use most of
+  the existing code as possible
+
+
+### Low-level implementation of high-level concepts
+
+The points here (tagging/unboxing, and big/small values) are actually
+tasks that are also performed by cmmgen (because they are properties of the
+diffrecnes between representtion between middle-end and backend).
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/docs/to_cmm.md]: fixed

**Pull request merged**